### PR TITLE
OCPBUGS-4963: Enable base nodeip-configuration for vsphere upi

### PR DESCRIPTION
**Pull request by cybertron**
When the base nodeip-configuration service was added for UPI, the
vsphere platform was missed because it has a unique UPI configuration.
Instead of using platform None, it uses platform vsphere but does
not set the VIP fields. As a result, neither nodeip-configuration
service was enabled on that platform and multiple nic behavior is
problematic.

Because the _base nodeip-configuration service is overwritten by the
IPI version in the on-prem directory, we need a separate service for
VSphere UPI that can be enabled when platform is "VSphere" and the
API VIPs field is not set. This is the inverse logic to the one in
the on-prem directory, so only one will ever be enabled at a time.

**- What I did**

**- How to verify it**

**- Description for the changelog**
Enabled the nodeip-configuration service for the VSphere UPI platform.

### CODE REVIEW

**Comment by rioliu-rh** [comment]
/hold
@cybertron hi Ben, I installed a vsphere cluster with this pr code, found that it does not work, the `platformStatus.type` is `VSphere`, it starts with uppercase, can you confirm

```
oc get infrastructure/cluster -o yaml |yq -y '.status'
apiServerInternalURI: https://api-int.rioliu-1216a.qe.devcluster.openshift.com:6443
apiServerURL: https://api.rioliu-1216a.qe.devcluster.openshift.com:6443
controlPlaneTopology: HighlyAvailable
etcdDiscoveryDomain: ''
infrastructureName: rioliu-1216a-p7lm7
infrastructureTopology: HighlyAvailable
platform: VSphere
platformStatus:
  type: VSphere
```
check the service status of `nodeip-configuration`, it is disabled.
```
oc debug --quiet node/rioliu-1216a-p7lm7-compute-0 -- chroot /host systemctl status nodeip-configuration.service
● nodeip-configuration.service - Writes IP address configuration so that kubelet and crio services select a valid node IP
   Loaded: loaded (/etc/systemd/system/nodeip-configuration.service; disabled; vendor preset: disabled)
   Active: inactive (dead)
error: non-zero exit code from debug container
```

**Comment by cybertron** [comment]
Oh, whoops! I looked up the name we return here: https://github.com/openshift/machine-config-operator/blob/5b8bd66b7fdd07df29a766c43d4fbeec177e25d0/pkg/operator/render.go#L235 but then used a different field in the check. Thanks for catching it!

**Comment by rioliu-rh** [comment]
> Oh, whoops! I looked up the name we return here:
> 
> https://github.com/openshift/machine-config-operator/blob/5b8bd66b7fdd07df29a766c43d4fbeec177e25d0/pkg/operator/render.go#L235
> 
> but then used a different field in the check. Thanks for catching it!

Hi Ben (@cybertron ) I setup a new cluster with latest code, the service is still disabled. 
```
oc get infrastructure/cluster -o yaml |yq -y '.status'
apiServerInternalURI: https://api-int.rioliu-1219a.qe.devcluster.openshift.com:6443
apiServerURL: https://api.rioliu-1219a.qe.devcluster.openshift.com:6443
controlPlaneTopology: HighlyAvailable
etcdDiscoveryDomain: ''
infrastructureName: rioliu-1219a-dc6bw
infrastructureTopology: HighlyAvailable
platform: VSphere
platformStatus:
  type: VSphere

oc debug --quiet node/rioliu-1219a-dc6bw-compute-0 -- chroot /host systemctl status nodeip-configuration.service
● nodeip-configuration.service - Writes IP address configuration so that kubelet and crio services select a valid node IP
   Loaded: loaded (/etc/systemd/system/nodeip-configuration.service; disabled; vendor preset: disabled)
   Active: inactive (dead)
error: non-zero exit code from debug container
```

**Comment by cybertron** [comment]
Ah, I finally see what's happening here. Because platform is set to vsphere, nodeip-configuration is getting pulled in from the IPI version of the service. That is rightly disabled on VSphere since there are no VIPs.

It turns out this is tricky to deal with since there's no way to tell MCO to keep the _base version of the service just in this specific case. In order to avoid adding a bunch of platform-specific logic to the templating code, I'm just going to create a duplicate of the _base service in the vsphere directory which can be enabled for VSphere UPI deployments. If anyone has other suggestions on how to handle this I'm all ears though. I just don't see an elegant way to handle it.

**Comment by cybertron** [comment]
Looks like this is working now. I can see that the new service ran and set the kubelet node ip.

/test e2e-vsphere-upgrade
/test e2e-vsphere-upi

**Review by yuqi-zhang — APPROVED** [review]
Makes sense to me, but will let Rio confirm and un-hold once we qe-approve

**Comment by dcbw** [comment]
@rioliu-rh were you able to test this out and confirm?

**Comment by cybertron** [comment]
@rioliu-rh It would be nice to get this backported before 4.13 ships. Are you able to verify the latest version of this PR so we can get it merged in 4.14?

**Comment by rioliu-rh** [comment]
> @rioliu-rh It would be nice to get this backported before 4.13 ships. Are you able to verify the latest version of this PR so we can get it merged in 4.14?

working on it.

**Comment by rioliu-rh** [comment]
the cluster build via cluster-bot with param `vsphere,upi` does not have `APIServerInternalIPs`, so the service is still disabled

**Comment by cybertron** [comment]
> the cluster build via cluster-bot with param `vsphere,upi` does not have `APIServerInternalIPs`, so the service is still disabled, normally it works on ipi cluster infra is like blew

Sorry, I missed your reply. Note that the service name changed in the latest version of the PR. nodeip-configuration will still be disabled, but nodeip-configuration-vsphere-upi should be enabled and provides the same functionality.

/jira refresh

**Comment by rioliu-rh** [comment]
setup vsphere upi cluster via cluster-bot, check service status of `nodeip-configuration-vsphere-upi`, the service is enabled

```
oc get infrastructure/cluster -o yaml | yq -y '.status'
apiServerInternalURI: https://api-int.ci-ln-6qzxy0t-c1627.vmc-ci.devcluster.openshift.com:6443
apiServerURL: https://api.ci-ln-6qzxy0t-c1627.vmc-ci.devcluster.openshift.com:6443
controlPlaneTopology: HighlyAvailable
etcdDiscoveryDomain: ''
infrastructureName: ci-ln-6qzxy0t-c1627-d22nt
infrastructureTopology: HighlyAvailable
platform: VSphere
platformStatus:
  type: VSphere
```

```
debug node/compute-0 -- chroot /host systemctl status nodeip-configuration-vsphere-upi
○ nodeip-configuration-vsphere-upi.service - Writes IP address configuration so that kubelet and crio services select a valid node IP
     Loaded: loaded (/etc/systemd/system/nodeip-configuration-vsphere-upi.service; enabled; preset: disabled)
     Active: inactive (dead) since Wed 2023-03-29 01:08:26 UTC; 25min ago
   Main PID: 1345 (code=exited, status=0/SUCCESS)
        CPU: 271ms

Mar 29 01:08:26 compute-0 bash[1243]: time="2023-03-29T01:08:26Z" level=info msg="Opening path /run/nodeip-configuration/primary-ip"
Mar 29 01:08:26 compute-0 bash[1243]: time="2023-03-29T01:08:26Z" level=info msg="Writing path /run/nodeip-configuration/primary-ip with content 192.168.52.7"
Mar 29 01:08:26 compute-0 bash[1243]: time="2023-03-29T01:08:26Z" level=info msg="Opening path /run/nodeip-configuration/ipv4"
Mar 29 01:08:26 compute-0 nice_margulis[1322]: time="2023-03-29T01:08:26Z" level=info msg="Opening path /run/nodeip-configuration/ipv4"
Mar 29 01:08:26 compute-0 bash[1243]: time="2023-03-29T01:08:26Z" level=info msg="Writing path /run/nodeip-configuration/ipv4 with content 192.168.52.7"
Mar 29 01:08:26 compute-0 nice_margulis[1322]: time="2023-03-29T01:08:26Z" level=info msg="Writing path /run/nodeip-configuration/ipv4 with content 192.168.52.7"
Mar 29 01:08:26 compute-0 podman[1332]: 2023-03-29 01:08:26.732695183 +0000 UTC m=+0.034348362 container died 0d5b6c792ba219ef7eb5d2875b4461884db6df9a4ddb26a1b44d4ea544138260 (image=registry.apps.build01-us-west-2.vmc.ci.openshift.org/ci-ln-6qzxy0t/stable@sha256:40cda5fb6d1b85aac5696cd4ae0edceb5953c9b767b4455c0da98a5a65378be5, name=nice_margulis, distribution-scope=public, io.openshift.build.source-context-dir=, io.openshift.ci.from.base-os=sha256:b5d707159189d8a395cba3e532d64bdf4cb6d1f2781c927d78ae2d49c371f157, vendor=Red Hat, Inc., maintainer=Antoni Segura Puimedon <antoni@redhat.com>, build-date=2023-02-15T04:02:00, io.openshift.expose-services=, version=8.7, description=The Universal Base Image is designed and engineered to be the base layer for all of your containerized applications, middleware and utilities. This base image is freely redistributable, but Red Hat only supports Red Hat technologies through subscriptions for Red Hat products. This image is maintained by Red Hat and updated regularly., com.redhat.component=ubi8-container, vcs-url=https://github.com/openshift/baremetal-runtimecfg, io.openshift.build.name=, io.openshift.build.namespace=, url=https://access.redhat.com/containers/#/registry.access.redhat.com/ubi8/images/8.7-1090, io.openshift.build.commit.date=, io.openshift.build.commit.message=, summary=Provides the latest release of Red Hat Universal Base Image 8., io.k8s.display-name=baremetal-runtimecfg, release=1090, architecture=x86_64, io.openshift.build.commit.author=, io.openshift.build.source-location=https://github.com/openshift/baremetal-runtimecfg, io.openshift.build.commit.ref=master, vcs-ref=5847e8945c4997c5257c9f4f74cde0800e555866, vcs-type=git, io.openshift.build.commit.id=5847e8945c4997c5257c9f4f74cde0800e555866, io.k8s.description=Retrieves Node and Cluster information for baremetal network config, io.buildah.version=1.29.0, com.redhat.license_terms=https://www.redhat.com/en/about/red-hat-end-user-license-agreements#UBI, io.openshift.ci.from.ubi8=sha256:b7e686e30346e9ace664fa09c0275262f8b9a443ed56d22165a0e201f6488c13, name=ubi8, io.openshift.tags=base rhel8)
Mar 29 01:08:26 compute-0 podman[1332]: 2023-03-29 01:08:26.754159494 +0000 UTC m=+0.055812670 container remove 0d5b6c792ba219ef7eb5d2875b4461884db6df9a4ddb26a1b44d4ea544138260 (image=registry.apps.build01-us-west-2.vmc.ci.openshift.org/ci-ln-6qzxy0t/stable@sha256:40cda5fb6d1b85aac5696cd4ae0edceb5953c9b767b4455c0da98a5a65378be5, name=nice_margulis, io.openshift.build.source-context-dir=, vcs-ref=5847e8945c4997c5257c9f4f74cde0800e555866, io.openshift.build.source-location=https://github.com/openshift/baremetal-runtimecfg, summary=Provides the latest release of Red Hat Universal Base Image 8., io.openshift.tags=base rhel8, io.k8s.display-name=baremetal-runtimecfg, io.openshift.build.namespace=, com.redhat.component=ubi8-container, io.openshift.build.commit.id=5847e8945c4997c5257c9f4f74cde0800e555866, url=https://access.redhat.com/containers/#/registry.access.redhat.com/ubi8/images/8.7-1090, vcs-type=git, vcs-url=https://github.com/openshift/baremetal-runtimecfg, io.openshift.build.commit.message=, vendor=Red Hat, Inc., io.openshift.ci.from.base-os=sha256:b5d707159189d8a395cba3e532d64bdf4cb6d1f2781c927d78ae2d49c371f157, description=The Universal Base Image is designed and engineered to be the base layer for all of your containerized applications, middleware and utilities. This base image is freely redistributable, but Red Hat only supports Red Hat technologies through subscriptions for Red Hat products. This image is maintained by Red Hat and updated regularly., io.openshift.build.name=, version=8.7, distribution-scope=public, io.openshift.build.commit.ref=master, build-date=2023-02-15T04:02:00, name=ubi8, io.openshift.expose-services=, io.openshift.build.commit.date=, io.openshift.ci.from.ubi8=sha256:b7e686e30346e9ace664fa09c0275262f8b9a443ed56d22165a0e201f6488c13, io.buildah.version=1.29.0, release=1090, architecture=x86_64, io.k8s.description=Retrieves Node and Cluster information for baremetal network config, io.openshift.build.commit.author=, com.redhat.license_terms=https://www.redhat.com/en/about/red-hat-end-user-license-agreements#UBI, maintainer=Antoni Segura Puimedon <antoni@redhat.com>)
Mar 29 01:08:26 compute-0 systemd[1]: nodeip-configuration-vsphere-upi.service: Deactivated successfully.
Mar 29 01:08:26 compute-0 systemd[1]: Finished Writes IP address configuration so that kubelet and crio services select a valid node IP.
```

**Pull request merged**
